### PR TITLE
VisIt: Add patch to include jpeg library in install.

### PR DESCRIPTION
**Pull request by brugger1**
When VisIt is run in parallel it needs the jpeg library to be installed in it's `lib` directory. This patch modifies the CMake related files to accomplish this.

I tested this building VisIt 3.2.2 and 3.3.0 on crusher.olcf.ornl.gov.

### CODE REVIEW

**Comment by alalazo** [comment]
Waiting for @cyrush approval also on this one. Feel free to ping me to merge.

**Review by cyrush — DISMISSED** [review]

**Comment by cyrush** [comment]
@alalazo  approved

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by brugger1** [comment]
@spackbot run pipeline

**Comment by brugger1** [comment]
@alalazo The vis-sdk test failed and the log file got truncated so it isn't possible to see why it failed. I tried to rerun the tests with the magic command to the spack bot, but nothing has happened.

**Comment by tgamblin** [comment]
@brugger1 looks like it is running now.

If it helps, you can reproduce the build exactly as it is in CI locally with this:

```
spack ci reproduce-build <URL of gitlab job>
```

That will pull the same binaries the PR used and try to build with your recipe.

**Comment by brugger1** [comment]
@tgamblin Thanks for the pointer on running the ci test manually. I did that and got the output below, which shows that there are some download failures, at a minimum for `visit3.2.2.tar.gz` from a mirror site and maybe a patch file from the PR. Clearly, the failure is before it even tries to build visit. The paths it is trying for the download don't quite make sense to me. I have no idea how to fix this.
Here is the actual output. Look for SpackWebError and NoSuchPatchError.
```
==> [2022-08-15-18:06:15.891254] Installing visit-3.2.2-zwgew65dvnqt56j4vfj3fqbm6qjpzgvb
==> [2022-08-15-18:06:15.891406] Searching for binary cache of visit-3.2.2-zwgew65dvnqt56j4vfj3fqbm6qjpzgvb
==> [2022-08-15-18:06:23.013147] No binary for visit-3.2.2-zwgew65dvnqt56j4vfj3fqbm6qjpzgvb found: installing from source
==> [2022-08-15-18:06:23.013725] Creating the installation directory /home/software/spack/[padded-to-512-chars]/morepadding/linux-ubuntu18.04-x86_64/gcc-7.5.0/visit-3.2.2-zwgew65dvnqt56j4vfj3fqbm6qjpzgvb
==> [2022-08-15-18:06:23.078985] '/usr/bin/git' '-C' '/spack' 'rev-parse' 'HEAD'
==> [2022-08-15-18:06:23.121236] Reading config file /builds/spack/spack/jobs_scratch_dir/reproduction/spack.yaml
==> [2022-08-15-18:06:23.232404] Using environment '/builds/spack/spack/jobs_scratch_dir/reproduction'
==> [2022-08-15-18:07:29.302252] [TOOLCHAIN] MATCH Gcc
==> [2022-08-15-18:07:29.317306] Adding compiler bin/ paths: /spack/lib/spack/env /spack/lib/spack/env/case-insensitive /spack/lib/spack/env/gcc
==> [2022-08-15-18:07:29.451290] No libraries found for util-macros
==> [2022-08-15-18:07:29.452532] No headers found for util-macros
==> [2022-08-15-18:07:29.495410] No libraries found for numactl
==> [2022-08-15-18:07:29.538123] No libraries found for tar
==> [2022-08-15-18:07:29.538941] No headers found for tar
==> [2022-08-15-18:07:29.716914] '/home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/linux-ubuntu18.04-x86_64/gcc-7.5.0/python-3.7.13-hujv6plqju42denlecadnovb7jl5xe6v/bin/python3.7' '-c' '
import json
from sysconfig import (
    get_config_vars,
    get_config_h_filename,
    get_makefile_filename,
    get_paths,
)

config = get_config_vars()
config['"'"'config_h_filename'"'"'] = get_config_h_filename()
config['"'"'makefile_filename'"'"'] = get_makefile_filename()
config.update(get_paths())

print(json.dumps(config))
'
==> [2022-08-15-18:07:29.961810] No libraries found for sz
==> [2022-08-15-18:07:30.066579] No libraries found for glproto
==> [2022-08-15-18:07:30.083285] No libraries found for inputproto
==> [2022-08-15-18:07:30.096480] No libraries found for kbproto
==> [2022-08-15-18:07:30.115359] No libraries found for libpthread-stubs
==> [2022-08-15-18:07:30.116224] No headers found for libpthread-stubs
==> [2022-08-15-18:07:30.126561] No libraries found for xproto
==> [2022-08-15-18:07:30.138912] No libraries found for libxau
==> [2022-08-15-18:07:30.151663] No libraries found for libxdmcp
==> [2022-08-15-18:07:30.165493] No libraries found for xcb-proto
==> [2022-08-15-18:07:30.166186] No headers found for xcb-proto
==> [2022-08-15-18:07:30.183534] No libraries found for xextproto
==> [2022-08-15-18:07:30.200557] No libraries found for xtrans
==> [2022-08-15-18:07:30.313137] No libraries found for binutils
==> [2022-08-15-18:07:30.323115] '/home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/linux-ubuntu18.04-x86_64/gcc-7.5.0/llvm-12.0.1-fcc6wwjvimpjgnwouowsnddlbf32fxta/bin/llvm-config' '--libfiles' 'all'
==> [2022-08-15-18:07:30.832960] No libraries found for renderproto
==> [2022-08-15-18:07:30.852727] No libraries found for randrproto
==> [2022-08-15-18:07:30.876291] No libraries found for xrandr
==> [2022-08-15-18:07:30.876833] No headers found for xrandr
==> [2022-08-15-18:07:30.898090] No libraries found for mesa
==> [2022-08-15-18:07:31.022286] No libraries found for font-util
==> [2022-08-15-18:07:31.022938] No headers found for font-util
==> [2022-08-15-18:07:31.077252] No libraries found for pixman
==> [2022-08-15-18:07:31.121739] No libraries found for gobject-introspection
==> [2022-08-15-18:07:31.142765] No libraries found for icu4c
==> [2022-08-15-18:07:31.221007] No libraries found for xkbdata
==> [2022-08-15-18:07:31.221708] No headers found for xkbdata
==> [2022-08-15-18:07:31.265854] No libraries found for xcb-util-image
==> [2022-08-15-18:07:31.285374] No libraries found for xcb-util-keysyms
==> [2022-08-15-18:07:31.302197] No libraries found for xcb-util-renderutil
==> [2022-08-15-18:07:31.321357] No libraries found for xcb-util-wm
==> [2022-08-15-18:07:31.426972] No libraries found for qt
==> [2022-08-15-18:07:32.094351] No libraries found for glew
==> [2022-08-15-18:07:32.204831] No libraries found for vtk
==> [2022-08-15-18:07:32.365267] No libraries found for vtk-m
==> [2022-08-15-18:07:34.369685] A dependency has updated CPATH, this may lead pkg-config to assume that the package is part of the system includes and omit it when invoked with '--cflags'.
==> [2022-08-15-18:07:34.393664] '/usr/bin/g++' '-v' '/tmp/spack-implicit-link-infoib5kopg4/main.c' '-o' '/tmp/spack-implicit-link-infoib5kopg4/output'
==> [2022-08-15-18:07:34.680126] parsing implicit link info
==> [2022-08-15-18:07:34.681164] linker line:  /usr/lib/gcc/x86_64-linux-gnu/7/collect2 -plugin /usr/lib/gcc/x86_64-linux-gnu/7/liblto_plugin.so -plugin-opt=/usr/lib/gcc/x86_64-linux-gnu/7/lto-wrapper -plugin-opt=-fresolution=/tmp/cc3HkMxJ.res -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lgcc --build-id --eh-frame-hdr -m elf_x86_64 --hash-style=gnu --as-needed -dynamic-linker /lib64/ld-linux-x86-64.so.2 -pie -z now -z relro -o /tmp/spack-implicit-link-infoib5kopg4/output /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/Scrt1.o /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/7/crtbeginS.o -L/usr/lib/gcc/x86_64-linux-gnu/7 -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/7/../../.. /tmp/ccs6kqyn.o -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc /usr/lib/gcc/x86_64-linux-gnu/7/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu/crtn.o
==> [2022-08-15-18:07:34.681459] linkdir: /usr/lib/gcc/x86_64-linux-gnu/7
==> [2022-08-15-18:07:34.681603] linkdir: /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu
==> [2022-08-15-18:07:34.681695] linkdir: /usr/lib/gcc/x86_64-linux-gnu/7/../../../../lib
==> [2022-08-15-18:07:34.681794] linkdir: /lib/x86_64-linux-gnu
==> [2022-08-15-18:07:34.681910] linkdir: /lib/../lib
==> [2022-08-15-18:07:34.682009] linkdir: /usr/lib/x86_64-linux-gnu
==> [2022-08-15-18:07:34.682120] linkdir: /usr/lib/../lib
==> [2022-08-15-18:07:34.682251] linkdir: /usr/lib/gcc/x86_64-linux-gnu/7/../../..
==> [2022-08-15-18:07:34.682391] found raw link dirs: /usr/lib/gcc/x86_64-linux-gnu/7, /usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu, /usr/lib/gcc/x86_64-linux-gnu/7/../../../../lib, /lib/x86_64-linux-gnu, /lib/../lib, /usr/lib/x86_64-linux-gnu, /usr/lib/../lib, /usr/lib/gcc/x86_64-linux-gnu/7/../../..
==> [2022-08-15-18:07:34.682737] found link dirs: /usr/lib/gcc/x86_64-linux-gnu/7, /usr/lib/x86_64-linux-gnu, /usr/lib, /lib/x86_64-linux-gnu, /lib
==> [2022-08-15-18:07:34.725453] Fetching s3://spack-binaries-prs/pr32017_task/brugger1/2022_08_09_visit_jpeg_patch/_source-cache/archive/d1/d19ac24c622a3bc0a71bc9cd6e5c9860e43f39e3279672129278b6ebce8d0ead.tar.gz
==> [2022-08-15-18:07:35.763509] SpackWebError: Download failed: <urlopen error An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.>
==> [2022-08-15-18:07:35.772752] Fetching s3://spack-binaries-prs/pr32017_task/brugger1/2022_08_09_visit_jpeg_patch/visit/visit-3.2.2.tar.gz
==> [2022-08-15-18:07:37.631043] SpackWebError: Download failed: <urlopen error An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.>
==> [2022-08-15-18:07:37.641479] Fetching s3://spack-binaries/develop/data-vis-sdk/_source-cache/archive/d1/d19ac24c622a3bc0a71bc9cd6e5c9860e43f39e3279672129278b6ebce8d0ead.tar.gz
==> [2022-08-15-18:07:38.715950] SpackWebError: Download failed: <urlopen error An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.>
==> [2022-08-15-18:07:38.724978] Fetching s3://spack-binaries/develop/data-vis-sdk/visit/visit-3.2.2.tar.gz
==> [2022-08-15-18:07:39.808374] SpackWebError: Download failed: <urlopen error An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.>
==> [2022-08-15-18:07:39.819202] Checking existence of https://mirror.spack.io/_source-cache/archive/d1/d19ac24c622a3bc0a71bc9cd6e5c9860e43f39e3279672129278b6ebce8d0ead.tar.gz
==> [2022-08-15-18:07:40.740570] Fetching https://mirror.spack.io/_source-cache/archive/d1/d19ac24c622a3bc0a71bc9cd6e5c9860e43f39e3279672129278b6ebce8d0ead.tar.gz
==> [2022-08-15-18:07:56.087637] Fetching from s3://spack-binaries-prs/pr32017_task/brugger1/2022_08_09_visit_jpeg_patch/_source-cache/archive/d1/d19ac24c622a3bc0a71bc9cd6e5c9860e43f39e3279672129278b6ebce8d0ead.tar.gz failed.
==> [2022-08-15-18:07:56.088747] Fetching from s3://spack-binaries-prs/pr32017_task/brugger1/2022_08_09_visit_jpeg_patch/visit/visit-3.2.2.tar.gz failed.
==> [2022-08-15-18:07:56.089943] Fetching from s3://spack-binaries/develop/data-vis-sdk/_source-cache/archive/d1/d19ac24c622a3bc0a71bc9cd6e5c9860e43f39e3279672129278b6ebce8d0ead.tar.gz failed.
==> [2022-08-15-18:07:56.090478] Fetching from s3://spack-binaries/develop/data-vis-sdk/visit/visit-3.2.2.tar.gz failed.
==> [2022-08-15-18:07:57.469227] Error: NoSuchPatchError: Couldn't find patch for package builtin.visit with sha256: e26625fbec77381dd8a7daa6b4727189e2eabd0925645942154db9217615c300

/spack/lib/spack/spack/package_base.py:1534, in do_fetch:
       1531            self.stage.check()
       1532
       1533        self.stage.cache_local()
  >>   1534
       1535        for patch in self.spec.patches:
       1536            patch.fetch()
       1537            if patch.stage:


Traceback (most recent call last):
  File "/spack/lib/spack/spack/build_environment.py", line 1061, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
  File "/spack/lib/spack/spack/installer.py", line 1987, in build_process
    return installer.run()
  File "/spack/lib/spack/spack/installer.py", line 1824, in run
    self.pkg.do_patch()
  File "/spack/lib/spack/spack/package_base.py", line 1563, in do_patch
    self.do_stage()
  File "/spack/lib/spack/spack/package_base.py", line 1548, in do_stage
    self.do_fetch(mirror_only)
  File "/spack/lib/spack/spack/package_base.py", line 1535, in do_fetch
    for patch in self.spec.patches:
  File "/spack/lib/spack/spack/spec.py", line 3755, in patches
    patch = index.patch_for_package(sha256, pkg_cls)
  File "/spack/lib/spack/spack/patch.py", line 361, in patch_for_package
    "Couldn't find patch for package %s with sha256: %s" % (pkg.fullname, sha256)
spack.patch.NoSuchPatchError: Couldn't find patch for package builtin.visit with sha256: e26625fbec77381dd8a7daa6b4727189e2eabd0925645942154db9217615c300
==> [2022-08-15-18:07:57.556998] Flagging visit-3.2.2-zwgew65dvnqt56j4vfj3fqbm6qjpzgvb as failed: NoSuchPatchError: Couldn't find patch for package builtin.visit with sha256: e26625fbec77381dd8a7daa6b4727189e2eabd0925645942154db9217615c300
==> [2022-08-15-18:07:57.643108] Downgrading to a read lock on visit-3.2.2-zwgew65dvnqt56j4vfj3fqbm6qjpzgvb with timeout 1e-09
==> [2022-08-15-18:07:57.643661] visit-3.2.2-zwgew65dvnqt56j4vfj3fqbm6qjpzgvb is now read locked
```

**Comment by scottwittenburg** [comment]
@brugger1 We don't store patches in those s3 mirrors mentioned in the log above (`s3://spack-binaries-prs` or `s3://spack-binaries`), so the fact that spack didn't find any patches there isn't surprising to me.  It's just looking in all the mirrors it knows about (those s3 mirrors are places where spack might find binaries to install).  I don't fully understand the mechanisms by which source patches are hashed, stored, and retrieved, but to me the surprising line from the log above is that it didn't find the source patch here:

```
==> [2022-08-15-18:07:40.740570] Fetching https://mirror.spack.io/_source-cache/archive/d1/d19ac24c622a3bc0a71bc9cd6e5c9860e43f39e3279672129278b6ebce8d0ead.tar.gz
```

To me that looks likely to be the place where the patch could actually be found, maybe @alalazo has some insight, but I don't know what's going on there.

**Comment by scottwittenburg** [comment]
Maybe that error you saw was temporary @brugger1?  I just ran the reproducer and encountered the what appears to be the same error I pasted in slack:

```
1 error found in build log:
     124    --   QWT found
     125    -- PySide NOT found
     126    -- Looking for ZLIB
     127    --   Found library z in /home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path
            _placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/linux-ubuntu18.04-x86_64/gcc-7.5.0/zlib-1.2.12-m4fqokrwbprfci4rptvgevnklpdo3dhd/lib
     128    --   ZLIB found
     129    -- Looking for JPEG
  >> 130    CMake Error at CMake/SetUpThirdParty.cmake:136 (MESSAGE):
     131         None of library directories for JPEG (/home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeho
            lder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/linux-ubuntu18.04-x86_64/gcc-7.5.0/libjpeg-turbo-2.1.3-d4sxegz4sgwkb4comrxbyop6bsgwsr62/lib64) exi
            st.
     132    
     133      **
     134    
     135      **
     136    Call Stack (most recent call first):
```

**Comment by brugger1** [comment]
I understand that error message. So, yes, the failure I got must have transitory. Thanks a lot!

**Comment by brugger1** [comment]
@alalazo I fixed the ci/gitlab-ci failure and now I'm getting a failure with readthedocs. I had the same issue with another PR and updating the branch fixed that. Of course updating the branch brings in a lot of changes from develop and makes it look like I lot of things were changed, when in fact only a few were. Should I update the branch or just do a new PR off the current develop?

**Comment by brugger1** [comment]
Closing this PR and opening a new one.

**Pull request closed without merging**
